src/Plugin/Field/FieldFormatter/GroupSubscribeFormatter.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+namespace Drupal\og\Plugin\Field\FieldFormatter;
+
+use Drupal\Core\Field\FieldItemListInterface;
+use Drupal\Core\Field\FormatterBase;
+use Drupal\Core\Routing\RedirectDestinationTrait;
+use Drupal\Core\Url;
+use Drupal\og\Og;
+use Drupal\og\OgMembershipInterface;
+use Drupal\user\Entity\User;
+use Drupal\user\EntityOwnerInterface;
+
+/**
+ * Plugin implementation for the OG subscribe formatter.
+ *
+ * @FieldFormatter(
+ *   id = "og_group_subscribe",
+ *   label = @Translation("OG Group subscribe"),
+ *   description = @Translation("Display OG Group subscribe and un-subscribe links."),
+ *   field_types = {
+ *     "og_group"
+ *   }
+ * )
+ */
+class GroupSubscribeFormatter extends FormatterBase {
+
+  use RedirectDestinationTrait;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function viewElements(FieldItemListInterface $items, $langcode) {
+    $elements = [];
+
+    // We cannot use the field cache, as the formatter changes according to the
+    // user. Furthermore, this is not an expensive check, so we remove the cache
+    // entirely.
+    $elements['#cache']['max-age'] = 0;
+
+    $group = $items->getEntity();
+    $entity_type_id = $group->getEntityTypeId();
+
+    $user = User::load(\Drupal::currentUser()->id());
+    if (($group instanceof EntityOwnerInterface) && ($group->getOwnerId() == $user->id())) {
+      // User is the group manager.
+      $elements[0] = [
+        '#type' => 'html_tag',
+        '#tag' => 'span',
+        '#attributes' => [
+          'title' => $this->t('You are the group manager'),
+          'class' => ['group', 'manager'],
+        ],
+        '#value' => $this->t('You are the group manager'),
+      ];
+
+      return $elements;
+    }
+
+    /** @var \Drupal\og\OgAccessInterface $og_access */
+    $og_access = \Drupal::service('og.access');
+
+    if (Og::isMemberBlocked($group, $user)) {
+      // If user is blocked, they should not be able to apply for
+      // membership.
+      return $elements;
+    }
+
+    if (Og::isMember($group, $user, [OgMembershipInterface::STATE_ACTIVE, OgMembershipInterface::STATE_PENDING])) {
+      $link['title'] = $this->t('Unsubscribe from group');
+      $link['url'] = Url::fromRoute('og.unsubscribe', ['entity_type_id' => $entity_type_id, 'entity_id' => $group->id()]);
```

Comment:
This is triggering an exception for me in my project:

> Symfony\Component\Routing\Exception\InvalidParameterException: Parameter "entity_id" for route "og.unsubscribe" must match "[^/]++" ("http://placeHolder/d5aeb410-3a30-4f17-a0da-05cf507bafd1" given) to generate a corresponding URL. in Drupal\Core\Routing\UrlGenerator->doGenerate() (line 193 of core/lib/Drupal/Core/Routing/UrlGenerator.php)

We are using entities that are conforming to the RDF specification (semantic web) and our IDs are URLs that contain forward slashes.

In our own code base we have an alternative way to subscribe and unsubscribe from groups. For us it works since we know the entity type in advance and can specify it to automatically upcast the entity ID to a full entity in the route. Here is our route description:

```
collection.leave_confirm_form:
  path: '/rdf_entity/{rdf_entity}/leave'
  defaults:
    _form: '\Drupal\collection\Form\LeaveCollectionConfirmForm'
    _title: 'Leave collection'
  requirements:
    _custom_access: '\Drupal\collection\Form\LeaveCollectionConfirmForm::access'
  options:
    parameters:
      rdf_entity:
        type: entity:rdf_entity
```

This is not applicable directly in OG, since the entity type is dynamic and depends on the `entity_type_id` parameter in the route. I'm not sure how this can be solved.

Pinging my colleague @sandervd who is very knowledgeable about these things.